keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1461 @@
+
+In order to mitigate this risk currently the time of last failure is recorded
+in-memory (globally for all pods within a Job). And a new failed pod may bump
+it only until it is added to the `uncountedTerminatedPods` structure.
+
+However, tracking the last failure time per index might be costly for memory
+consumption (see [Expotential backoff delay with in-memory tracking](#expotential-backoff-delay-with-in-memory-tracking)).
+
+Thus, in order to mitigate this risk we propose to compute the finish time for
+a pod as the first available value of the following (avoiding the ever-increasing
+fallback to `now`):
+1. max `finishAt` of all containers, if specified for all containers
+2. `LastTransitionTime` for the `Ready=False` condition
+3. `deletionTimestamp` - `deletionGracePeriodSeconds` if `deletionTimestamp` is set
+
+Here (3.) is used to mark the moment of deletion which is used to approximate
+the current behavior. (2.) is used when Kubelet loses track of one of its containers,
+the `Ready=False` condition is set by Kubelet when transitioning a pod to `Failed`
+phase: https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/kubelet/status/status_manager.go#L1060-L1068.
+When none of the above conditions is satisfied to compute the finish time we
+fallback to the pod's creation time.
+
+This fix can be considered a preparatory PR before the KEP, as to some extent
+is solves the preexisting issue.
+
+### Too fast Job status updates
+
+In this KEP the Job controller needs to keep updating the new status field
+`.status.failedIndexes` to reflect the current status of the Job. This can raise
+concerns of overwhelming the API server with status updates.
+
+First, observe that the new field does not entail additional Job status updates.
+When a pod terminates (either failure or success), it triggers Job status update
+to increment the `status.failed` or `.status.succeeded` counter fields. These
+updates are also used to update the pre-existing `status.completedIndexes`
+field, and the new `status.failedIndexes` field.
+
+Second, in order to mitigate this risk there is already a mechanism present in
+the Job controller, to bulk Job status updates per Job.
+
+The way the mechanism works is that Job controller maintains a queue of `syncJob`
+invocations per job
+(see [in code](https://github.com/kubernetes/kubernetes/blob/72a3990728b2a8979effb37b9800beb3117349f6/pkg/controller/job/job_controller.go#L118)).
+New items are added to the queue with a delay (1s for pod events, such as:
+delete, add, update). The delay allows for deduplication of the sync per Job.
+
+One place to queue a new item in the queue, specific to this KEP, is when
+the expotential backoff delay hasn't elapsed for any index (allowing pod
+recreation), then we requeue the next Job status update. The delay is computed
+as minimum of all delays computed for all indexes requiring pod recreation,
+but not less that 1s.
```

Comment:
This is not strictly related to PRR, rather wearing my scalability hat now.
[I would also be happy to discuss it separately if preferred.]

My concern is purely code-related - you have this magic `immediate` parameter in:
https://github.com/kubernetes/kubernetes/blob/6195f96e56ee1e9f52986a0e768e22ca0d1949d6/pkg/controller/job/job_controller.go#L499
I have concerns:
1)  why we want to use a different separate backoff in those cases (that effectively starts from 0 instead of 1s)
2) why we don't do any batching when Job object is changing itself, but I think the new PR (that doesn't do batching only when the job generation changes) seems to address this part

Reply:
(1.) I agree. IIUC the idea of backoff is to throttle pod creation in case of consecutive failures. This is covered already here: https://github.com/kubernetes/kubernetes/blob/6195f96e56ee1e9f52986a0e768e22ca0d1949d6/pkg/controller/job/job_controller.go#L1415-L1419. However, currently it is also used for throttling in other places for non-obvious reasons. It also might be a left over after a relatively recent refactoring of how backoff works. I think this was always a little bit underspecified, +1 to clean it up at some point.  Also, I think it makes sense to modify the final backoff as `max(1s, computed backoff)`, but this is a detail. @alculquicondor @soltysh wdyt? Maybe we should open a separate Issue to clean up the backoff status?

(2.) yes, I don't think we want to batch (delay) job updates in case of spec updates. For example, spec updates are used to update suspend status, which could result in delayed job start.

Reply:
1. Backoff is used for errors from API calls (like conflicts). It is not used for events (pod creation, updates, etc). I guess a better name for the variable would be `useBackoff` or have separate functions altogether.
2. We want to be able to respond to changes to the spec ASAP (new parallelism, suspend, etc).

Reply:
> Just open a PR with your proposal of what a clean code looks like :)

We have one PR currently open for the fix + potentially small cleanup. I will consider another in the future, but some changes may not be just cosmetic, but also touching on semantics (so requiring discussion). Still, I suggest we move this discussion out of KEP to offline, under PRs or under Issues.

Reply:
For (1) - clearly backoff isn't used only for errors from API calls, because errors from API calls don't generate events and we have those in event handlers:
https://github.com/kubernetes/kubernetes/blob/6195f96e56ee1e9f52986a0e768e22ca0d1949d6/pkg/controller/job/job_controller.go#L318-L321

@mimowo - while opening PR is certainly good thing, can you open an issue to so that we won't forget about it?

For (2) - I agree.

Reply:
@wojtek-t sure, opened: https://github.com/kubernetes/kubernetes/issues/118527